File: Chapter11/scala/com/packtpub/spark/module_four/chapter_11/Deduplication.scala
package com.packtpub.spark.module_four.chapter_11

import org.apache.spark.sql.types.{StringType, StructField, StructType}
import org.apache.spark.sql.{Row, SparkSession}

object Deduplication {

  def main(args: Array[String]): Unit = {

    // Build a SparkSession in Local Mode
    val spark = SparkSession
      .builder()
      .master("local[2]")
      .appName("My Spark App")
      .getOrCreate()

    val categorized_animals = Seq(Row("dog", "pet"),
                                  Row("cat", "pet"),
                                  Row("bear", "wild"),
                                  Row("cat", "pet"),
                                  Row("cat", "pet"))

    val schema = List(
      StructField("name", StringType, nullable = true),
      StructField("category", StringType, nullable = true)
    )

    val animalDataRDD = spark.sparkContext.parallelize(categorized_animals)

    val animalData = spark.createDataFrame(animalDataRDD, StructType(schema))

    animalData.show()
    val deduped = animalData.dropDuplicates()
    deduped.show()

  }

}

File: Chapter11/python/Deduplication.py
from pyspark.sql import SparkSession
from pyspark.sql import Row
from pyspark.sql import SQLContext


# Function to read data in RDD
def read(animal): print(animal['name'], animal['category'])


# Create a Spark Session
spark = SparkSession\
    .builder\
    .appName("My Spark App")\
    .master("local[2]")\
    .getOrCreate()

sc = spark.sparkContext
sqlContext = SQLContext(sc)

# create a sample set of data as an RDD
categorized_animals = [("dog", "pet"), ("cat", "pet"), ("bear", "wild"), ("cat", "pet"), ("cat", "pet")]
animalDataRDD = sc.parallelize(categorized_animals)
animals = animalDataRDD.map(lambda x: Row(name=x[0], category=x[1]))
animalsDF = sqlContext.createDataFrame(animals)

animalsDF.foreach(read)

deduplicated = animalsDF.dropDuplicates()
deduplicated.foreach(read)
